packages/plugin-ext/src/main/browser/editor/untitled-resource.ts

```diff
@@ -14,39 +14,91 @@
+    protected readonly fileSystemWatcher: FileSystemWatcher;
+
+    @inject(OpenerService)
+    protected readonly openerService: OpenerService;
+
+    async resolve(uri: URI): Promise<UntitledResource> {
+        if (uri.scheme !== Schemes.UNTITLED) {
+            throw new Error('The given uri is not untitled file uri: ' + uri);
+        } else {
+            const untitledResource = resources.get(uri.toString());
+            if (!untitledResource) {
+                return createUntitledResource('', '', uri, this.fileSystem, this.fileSystemWatcher);
+            } else {
+                return untitledResource;
+            }
         }
-        throw new Error(`scheme ${uri.scheme} is not '${Schemes.UNTITLED}'`);
     }
 }
 
 export class UntitledResource implements Resource {
+    private fileResource: FileResource;
 
-    constructor(public uri: URI, private content?: string) {
+    constructor(public uri: URI, private content?: string, private fileSystem?: FileSystem, private fileSystemWatcher?: FileSystemWatcher) {
         resources.set(this.uri.toString(), this);
     }
 
-    readContents(options?: { encoding?: string | undefined; } | undefined): Promise<string> {
-        return Promise.resolve(this.content ? this.content : '');
-    }
-
     dispose(): void {
         resources.delete(this.uri.toString());
     }
+
+    async readContents(options?: { encoding?: string | undefined; } | undefined): Promise<string> {
+        if (this.content) {
+            return Promise.resolve(this.content);
+        } else if (this.fileSystem && await this.fileSystem.exists(this.uri.toString())) {
+            try {
+                const { stat, content } = await this.fileSystem.resolveContent(this.uri.toString(), options);
+                console.debug('The file : ' + this.uri.toString() + ' stat is :' + stat);
+                return content;
+            } catch (e) {
+                throw e;
+            }
+        } else if (this.fileResource) {
+            return this.fileResource.readContents(options);
+        } else {
+            return Promise.resolve('');
+        }
+    }
+
+    async saveContents(content: string, options?: { encoding?: string, overwriteEncoding?: string }): Promise<void> {
+        if (this.fileSystem && this.fileSystemWatcher) {
+            this.fileResource = new FileResource(this.uri, this.fileSystem, this.fileSystemWatcher);
+            await this.fileResource.init();
+            this.fileResource.saveContents(content, options);
+        }
+    }
+
+    async saveContentChanges(changes: TextDocumentContentChangeEvent[], options?: { encoding?: string, overwriteEncoding?: string }): Promise<void> {
+        if (this.fileResource) {
+            this.fileResource.saveContentChanges(changes, options);
+        }
+    }
 }
 
-export function createUntitledResource(content?: string, language?: string): UntitledResource {
+export async function createUntitledResource(content?: string, language?: string,
```

Comment:
It should be a method of `UntitledResourceResolver`. all arguments should be mandatory, `UntitledResource` should be refactored with an assumption that `content` is always given

Reply:
But what about this call https://github.com/eclipse-theia/theia/blob/master/packages/plugin-ext/src/main/browser/documents-main.ts#L172 tom the 'createUntitledResource' method ?